src/rt/sched/cown.h

```diff
@@ -647,23 +701,64 @@ namespace verona::rt
 
       Scheduler::local()->message_body = &body;
 
+      // The message `m` will be deallocated when the next scheduler thread
+      // picks up a message, so wait until after all the possible uses of `m` to
+      // reschedule this cown.
+      if (!schedule_after_behaviour)
+        schedule();
+
       // Run the behaviour.
       body.behaviour->f();
 
       for (size_t i = 0; i < body.count; i++)
       {
-        if (body.cowns[i])
-          Cown::release(alloc, body.cowns[i]);
+        if (body.requests[i].cown)
+          Cown::release(alloc, body.requests[i].cown);
       }
 
       Logging::cout() << "MultiMessage " << m << " completed and running on "
                       << this << Logging::endl;
 
       // Free the body and the behaviour.
       alloc.dealloc(body.behaviour, body.behaviour->size());
-      alloc.dealloc<sizeof(MultiMessage::MultiMessageBody)>(m->get_body());
 
-      return true;
+      //  Reschedule the writeable cowns (read-only cowns are not unscheduled
+      //  for a behaviour).
+      for (size_t s = 0; s < body.count; s++)
+      {
+        Request request = body.requests[s];
+        Cown* cown = request.cown;
+        if (cown)
+        {
+          if (request.mode == AccessMode::WRITE && cown != this)
+            cown->schedule();
+          else if (request.mode == AccessMode::READ)
+          {
+            // If this read is the last read of a cown, and that cown's next
+            // message requires writing, clear the flag and either:
+            //   - schedule the cown if it is not this cown, or
+            //   - continue processing this cown's message on this scheduler
+            //   thread
+            //     (overwriting the previous decision to stop processing this
+            //     cown).
+            size_t rc = cown->read_ref_count.fetch_sub(2);
```

Comment:
Just to check my understanding.  If `rc` is made `0` by this subtract, it is fine.  It may or not be scheduled, it is just up to if it has been marked sleeping.

Reply:
Yes, if `cown->read_ref_count` hits `0` here it only signals that there are no more reads happening on this cown right now, there may be nothing left in its queue to process, or we may not have processed the next mesage yet.